pkg/agent/billing/billing.go

```diff
@@ -146,31 +155,44 @@ func (s *metricsState) collect(conf *Config, store VMStoreForNode, metrics PromM
 	vmsOnThisNode := store.ListIndexed(func(i *VMNodeIndex) []*vmapi.VirtualMachine {
 		return i.List()
 	})
+	logger.Info(fmt.Sprintf("num_vms: %d", len(vmsOnThisNode)))
 	for _, vm := range vmsOnThisNode {
 		endpointID, isEndpoint := vm.Labels[EndpointLabel]
 		metricsBatch.inc(isEndpointFlag(isEndpoint), autoscalingEnabledFlag(api.HasAutoscalingEnabled(vm)), vm.Status.Phase)
-		if !isEndpoint {
-			// we're only reporting metrics for VMs with endpoint IDs, and this VM doesn't have one
-			continue
-		}
+		// if !isEndpoint {
+		// 	// we're only reporting metrics for VMs with endpoint IDs, and this VM doesn't have one
+		// 	continue
+		// }
 
 		if !vm.Status.Phase.IsAlive() || vm.Status.CPUs == nil {
+			logger.Info("Skipping gathering metrics for VM")
 			continue
 		}
 
+		byteCounts, err := vm.GetNetworkUsage()
+		if err != nil {
+			return err
+		}
```

Comment:
Some context: These lines are executed once for each billed VM, every `config.CollectEverySeconds`.

My primary concern here is that we might block the primary "collection" thread (which generates the "source of truth" for billing events) with network requests. Given that sometimes network requests can sometimes take a long time connecting, these at least should be isolated from the "collection" thread, and _maybe_ in individual goroutines.

Reply:
If we're just connecting to localhost I don't _think_ it'll be an issue, but I'll let you be the judge. How would you recommend isolating it? Would we have a separate goroutine that polls periodically and pushes the metrics into a channel that the main thread then reads from?

Reply:
It's unfortunately quite easy for network requests to indefinitely stall if the host we're trying to connect to doesn't exist (or, no longer exists).

I would probably use a worker pool to execute these requests, with some reasonably short timeout (e.g. 1s-5s). Realistically, it's _probably_ ok to make a new goroutine on each request, so worker pool is not strictly required (and may actually be worse - lots of resources doing nothing) -- the important piece is having some reasonable upper bound on the number of concurrent requests.

_Ideally_ we'd also have some metric acting as a counter for the number of requests and their outcomes (similar-ish to the existing metric we have for sending billing events).